### PR TITLE
Preallocate cloning DV

### PR DESCRIPTION
**Pull request by tomob**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Added support for Cloning DataVolumes.
```

### CODE REVIEW

File: cmd/cdi-cloner/clone-source.go [hunk 1]
```diff
@@ -199,10 +204,11 @@ func main() {
 	serverCert := []byte(getEnvVarOrDie("SERVER_CA_CERT"))
 
 	url := getEnvVarOrDie("UPLOAD_URL")
+	preallocation, _ := strconv.ParseBool(getEnvVarOrDie(common.Preallocation)) // False is default in case of error
```
Comment (ShellyKa13) [inline cmd/cdi-cloner/clone-source.go:207]: maybe if there was an error and it uses false as default should print some warning or something to explain that the default  false is used.

**Review by ShellyKa13 — COMMENTED** [review]
/lgtm

File: cmd/cdi-cloner/clone-source.go [hunk 2]
```diff
@@ -143,8 +144,12 @@ func validateMount() {
 	}
 }
 
-func newTarReader() (io.ReadCloser, error) {
-	cmd := exec.Command("/usr/bin/tar", "Scv", ".")
+func newTarReader(preallocation bool) (io.ReadCloser, error) {
+	args := "cv"
+	if !preallocation {
```
Comment (brybacki) [inline cmd/cdi-cloner/clone-source.go:149]: a very short comment that preallocation does not play well with sparse tar files

File: tests/cloner_test.go [hunk 3]
```diff
@@ -1135,6 +1135,90 @@ var _ = Describe("all clone tests", func() {
...
+			Expect(len(crList.Items)).To(Equal(1))
+
+			cdiCrSpec = crList.Items[0].Spec.DeepCopy()
+			cdiCr = crList.Items[0]
+
+			By("[BeforeEach] Forcing Host Assisted cloning")
+			var cloneStrategy cdiv1.CDICloneStrategy = cdiv1.CloneStrategyHostAssisted
+			cdiCr.Spec.CloneStrategyOverride = &cloneStrategy
+			_, err = f.CdiClient.CdiV1beta1().CDIs().Update(context.TODO(), &cdiCr, metav1.UpdateOptions{})
+			Expect(err).ToNot(HaveOccurred())
```
Comment (awels) [inline tests/cloner_test.go:1160]: You probably want to ensure that the clone strategy is set in a fresh object before going onto the test, sometimes it takes a few seconds to update, and during that time you don't want your test to run.

File: tests/cloner_test.go [hunk 4]
```diff
@@ -1135,6 +1135,90 @@ var _ = Describe("all clone tests", func() {
...
+
+			By("[AfterEach] Restoring CDI CR spec to original state")
+			crList, err := f.CdiClient.CdiV1beta1().CDIs().List(context.TODO(), metav1.ListOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(len(crList.Items)).To(Equal(1))
+
+			newCdiCr := crList.Items[0]
+			newCdiCr.Spec = *cdiCrSpec
+			_, err = f.CdiClient.CdiV1beta1().CDIs().Update(context.TODO(), &newCdiCr, metav1.UpdateOptions{})
+			Expect(err).ToNot(HaveOccurred())
```
Comment (awels) [inline tests/cloner_test.go:1183]: Same here but reverse, make sure the original value was restored.

**Review by awels — COMMENTED** [review]
Besides making sure the cdi cr values are properly set during testing, looks good.

File: tests/cloner_test.go (hunk 3 above)
Comment (tomob) [reply tests/cloner_test.go:1160]: Done

File: tests/cloner_test.go (hunk 4 above)
Comment (tomob) [reply tests/cloner_test.go:1186]: Done

File: cmd/cdi-cloner/clone-source.go (hunk 2 above)
Comment (tomob) [reply cmd/cdi-cloner/clone-source.go:149]: Done

File: cmd/cdi-cloner/clone-source.go (hunk 1 above)
Comment (tomob) [reply cmd/cdi-cloner/clone-source.go]: Done

File: tests/utils/pvc.go [hunk 5]
```diff
@@ -256,3 +258,55 @@ func WaitPVCDeletedByUID(clientSet *kubernetes.Clientset, pvcSpec *k8sv1.Persist
...
+	err := wait.PollImmediate(pvcPollInterval, pvcCreateTime, func() (bool, error) {
+		var err error
+		cr, err := getCdiCR(clientSet)
+		if err != nil {
+			if apierrs.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+		if cr.Spec.CloneStrategyOverride != nil && cloneStrategy == *cr.Spec.CloneStrategyOverride {
```
Comment (arnongilboa) [inline tests/utils/pvc.go:284]: This func and the following one are similar except this check which can be passed as a func parameter, so a refactor will be nice.

**Review by awels — APPROVED** [review]
/approve

**Comment by awels** [comment]
@tomob looks like some test failures here?

**Comment by tomob** [comment]
> @tomob looks like some test failures here?

PR #1727 fixes them. This PR depends on it in a way.

**Pull request merged**